packages/engine/src/datastore/schema/field_spec/mod.rs

```diff
@@ -123,40 +132,47 @@ impl FieldKey {
             )));
         }
 
-        Ok(FieldKey(name.to_string()))
+        Ok(Self(name.to_string()))
     }
 
+    /// Create a new private or hidden `FieldKey`
+    ///
+    /// Builds a `FieldKey` from a given name, the [`FieldSource`], and the [`FieldScope`]. `scope`
+    /// must be either [`FieldScope::Private`] or [`FieldScope::Hidden`].
+    ///
+    /// # Errors
+    ///
+    /// - Returns [`Error`] if name starts with [`PRIVATE_PREFIX`] or [`HIDDEN_PREFIX`]
+    /// - Returns [`Error`] if `scope` is [`FieldScope::Agent`]
     #[inline]
     pub fn new_private_or_hidden_scoped(
         name: &str,
         source: &FieldSource,
         scope: &FieldScope,
-    ) -> Result<FieldKey> {
+    ) -> Result<Self> {
+        // TODO: do we want these checks to only be present on debug builds
         if name.starts_with(PRIVATE_PREFIX) || name.starts_with(HIDDEN_PREFIX) {
             return Err(Error::from(format!(
                 "Field names cannot start with the protected prefixes: ['{}', '{}']",
                 PRIVATE_PREFIX, HIDDEN_PREFIX
             )));
         }
 
-        let mut key = String::new();
-        match scope {
-            FieldScope::Private => {
-                key.push_str(PRIVATE_PREFIX);
-            }
-            FieldScope::Hidden => {
-                key.push_str(HIDDEN_PREFIX);
-            }
+        let scope_prefix = match scope {
+            FieldScope::Private => PRIVATE_PREFIX,
+            FieldScope::Hidden => HIDDEN_PREFIX,
             FieldScope::Agent => {
                 return Err(Error::from(
                     "Use new_agent_scoped to create a key with FieldScope::Agent",
                 ));
             }
-        }
-        key.push_str(&source.unique_id()?);
-        key.push('_');
-        key.push_str(name);
-        Ok(FieldKey(key))
+        };
+        Ok(Self(format!(
```

Comment:
I remember reading somewhere that for some reason `.push_str` is actually faster than format. I'm still happy with this change as we haven't done any proper optimisations in the new codebase yet and this feels like a micro-optimisation that's untested. But just thought I'd highlight

Reply:
Using `push_str()` is probably faster when `String` is allocated with `with_capacity`, otherwise I think `format!` should be at least as fast, but yeah, micro-optimisation, which doesn't matter.